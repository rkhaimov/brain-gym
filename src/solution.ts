function consequences(permutation: string, table: Record<string, 'P' | 'N'>) {
  for (let index = 0; index < permutation.length; index += 1) {
    if (permutation[index] === '1' && permutation[index + 1] === '1') {
      const result = trim(permutation.slice(0, index) + '00' + permutation.slice(index + 2));

      if (table[result] === 'P') {
        table[permutation] = 'N';

        return;
      }
    }

    if (permutation[index] === '1') {
      const result = trim(permutation.slice(0, index) + '0' + permutation.slice(index + 1));

      if (table[result] === 'P') {
        table[permutation] = 'N';

        return;
      }
    }
  }

  table[permutation] = 'P';

  function trim(target: string) {
    return target
      .replace(/0+/g, '0')
      .replace(/^0/g, '')
      .replace(/0$/g, '');
  }
}

function alltil(til: number) {
  const base = '1';
  const result: string[] = [base];

  let index = 0;
  while (result.length < til) {
    const permutations = ways(result[index]);

    result.push(...permutations);

    index += 1;
  }

  return result;
}

function ways(parent: string) {
  return [parent + '1', parent + '01'];
}
