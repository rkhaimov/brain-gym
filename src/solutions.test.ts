import { generate } from './solution';

test('generates outcomes for 10', () => {
  expect(generate(10)).toMatchSnapshot();
});

test('generates outcomes for 20', () => {
  expect(generate(20)).toMatchSnapshot();
});

console.log(
  Object.keys(generate(40))
    .map(binary => parseInt(binary, 2))
    .sort()
    .map(number => number.toString(2))
);
