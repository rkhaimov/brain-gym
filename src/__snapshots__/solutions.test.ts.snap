// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates outcomes for 10 1`] = `
Object {
  "1": "P",
  "1001": "N",
  "101": "N",
  "11": "N",
  "111": "N",
}
`;

exports[`generates outcomes for 20 1`] = `
Object {
  "1": "P",
  "10001": "N",
  "1001": "N",
  "10011": "N",
  "101": "N",
  "1011": "N",
  "11": "N",
  "1101": "N",
  "111": "N",
  "1111": "P",
}
`;
